File: apps/worker/src/github.ts
/**
 * GitHub App integration for the worker.
 *
 * When a GitHub App is configured (`GITHUB_APP_ID` + `GITHUB_APP_PRIVATE_KEY`),
 * this module mints short-lived **installation access tokens** and uses them to
 * (a) clone private repositories for a preview build and (b) report preview
 * status back onto the pull request as a commit status. All of it is optional
 * and env-gated: with no app configured the helpers report "unconfigured" and
 * the deploy path falls back to anonymous (public-repo) clones with no PR
 * status posting.
 *
 * Implemented with `jose` (RS256 app JWT) + global `fetch` — no `octokit`
 * dependency — to keep the worker's footprint small.
 *
 * @module
 */

import { importPKCS8, SignJWT } from "jose";

import type { WorkerConfig } from "./config.js";
import type { Logger } from "pino";

/** GitHub REST API base. */
const GITHUB_API = "https://api.github.com";

/** A minted installation token with its expiry. */
interface InstallationToken {
  token: string;
  expiresAt: number; // epoch ms
}

/**
 * Thin GitHub App client. Construct once per worker; it caches installation
 * tokens until shortly before they expire.
 */
export class GitHubApp {
  private readonly appId: string;
  private readonly privateKeyPem: string;
  private readonly logger: Logger;
  private readonly tokenCache = new Map<string, InstallationToken>();

  private constructor(appId: string, privateKeyPem: string, logger: Logger) {
    this.appId = appId;
    this.privateKeyPem = privateKeyPem;
    this.logger = logger;
  }

  /**
   * Build a {@link GitHubApp} from config, or `null` when the app is not
   * configured (so callers can cleanly fall back to the anonymous path).
   */
  static fromConfig(config: WorkerConfig, logger: Logger): GitHubApp | null {
    if (!config.GITHUB_APP_ID || !config.GITHUB_APP_PRIVATE_KEY) return null;
    // Env files often store the PEM with escaped newlines; normalize them.
    const pem = config.GITHUB_APP_PRIVATE_KEY.replace(/\\n/g, "\n");
    return new GitHubApp(config.GITHUB_APP_ID, pem, logger);
  }

  /** Sign a short-lived app JWT (RS256) used to mint installation tokens. */
  private async appJwt(): Promise<string> {
    const key = await importPKCS8(this.privateKeyPem, "RS256");
    // `iat` back-dated 60s to tolerate clock skew; GitHub caps `exp` at 10 min.
    const now = Math.floor(Date.now() / 1000) - 60;
    return new SignJWT({})
      .setProtectedHeader({ alg: "RS256" })
      .setIssuer(this.appId)
      .setIssuedAt(now)
      .setExpirationTime(now + 9 * 60)
      .sign(key);
  }

  /**
   * Get a (cached) installation access token for `installationId`. Tokens are
   * reused until 60s before expiry.
   */
  async installationToken(installationId: string): Promise<string> {
    const cached = this.tokenCache.get(installationId);
    if (cached && cached.expiresAt - Date.now() > 60_000) return cached.token;

    const jwt = await this.appJwt();
    const res = await fetch(
      `${GITHUB_API}/app/installations/${installationId}/access_tokens`,
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${jwt}`,
          Accept: "application/vnd.github+json",
          "User-Agent": "shipyard-worker",
        },
      },
    );
    if (!res.ok) {
      throw new Error(
        `GitHub App: failed to mint installation token (${res.status} ${res.statusText})`,
      );
    }
    const body = (await res.json()) as { token: string; expires_at: string };
    const token: InstallationToken = {
      token: body.token,
      expiresAt: new Date(body.expires_at).getTime(),
    };
    this.tokenCache.set(installationId, token);
    return token.token;
  }

  /**
   * Authenticated HTTPS clone URL for `repoFullName` using an installation
   * token, e.g. `https://x-access-token:<token>@github.com/acme/app.git`.
   */
  async cloneUrl(repoFullName: string, installationId: string): Promise<string> {
    const token = await this.installationToken(installationId);
    return `https://x-access-token:${token}@github.com/${repoFullName}.git`;
  }

  /**
   * Post a commit status onto the PR head SHA (best-effort; failures are logged
   * and swallowed so status reporting never fails a deploy).
   */
  async postCommitStatus(args: {
    repoFullName: string;
    installationId: string;
    sha: string;
    state: "pending" | "success" | "failure" | "error";
    targetUrl?: string;
    description?: string;
  }): Promise<void> {
    try {
      const token = await this.installationToken(args.installationId);
      const res = await fetch(
        `${GITHUB_API}/repos/${args.repoFullName}/statuses/${args.sha}`,
        {
          method: "POST",
          headers: {
            Authorization: `Bearer ${token}`,
            Accept: "application/vnd.github+json",
            "Content-Type": "application/json",
            "User-Agent": "shipyard-worker",
          },
          body: JSON.stringify({
            state: args.state,
            context: "shipyard/preview",
            ...(args.targetUrl ? { target_url: args.targetUrl } : {}),
            ...(args.description ? { description: args.description.slice(0, 140) } : {}),
          }),
        },
      );
      if (!res.ok) {
        this.logger.warn(
          { status: res.status, repo: args.repoFullName },
          "GitHub App: commit status post failed",
        );
      }
    } catch (err) {
      this.logger.warn({ err, repo: args.repoFullName }, "GitHub App: commit status error");
    }
  }
}

File: services/reviewbot/reviewbot/github.py
"""GitHub App integration: installation tokens + posting review comments.

Mirrors the TS worker's ``github.ts`` (RS256 App JWT → installation token →
REST calls) so the two services authenticate identically. All methods are
best-effort from the worker's perspective — a failed comment post must never
fail the review job.
"""

from __future__ import annotations

import logging
import time

import httpx
import jwt

logger = logging.getLogger(__name__)

_API = "https://api.github.com"
_TOKEN_SKEW_SECONDS = 60


class GitHubApp:
    """Minimal GitHub App client (installation-token auth)."""

    def __init__(self, app_id: str, private_key_pem: str) -> None:
        self._app_id = app_id
        self._private_key = private_key_pem
        #: installation id -> (token, expires_at_epoch)
        self._tokens: dict[str, tuple[str, float]] = {}

    @classmethod
    def from_config(cls, app_id: str | None, private_key: str | None) -> "GitHubApp | None":
        """Build the client when both credentials are present, else None."""
        if not app_id or not private_key:
            return None
        return cls(app_id, private_key)

    def _app_jwt(self) -> str:
        """Mint a short-lived RS256 JWT identifying the App itself."""
        now = int(time.time())
        payload = {"iat": now - 30, "exp": now + 9 * 60, "iss": self._app_id}
        return jwt.encode(payload, self._private_key, algorithm="RS256")

    def installation_token(self, installation_id: str) -> str:
        """Get (or refresh) an installation access token."""
        cached = self._tokens.get(installation_id)
        if cached and cached[1] - _TOKEN_SKEW_SECONDS > time.time():
            return cached[0]

        response = httpx.post(
            f"{_API}/app/installations/{installation_id}/access_tokens",
            headers={
                "Authorization": f"Bearer {self._app_jwt()}",
                "Accept": "application/vnd.github+json",
                "User-Agent": "shipyard-reviewbot",
            },
            timeout=30.0,
        )
        response.raise_for_status()
        data = response.json()
        # expires_at is ISO-8601; keep it simple with a fixed 55-minute window
        # (App tokens live 60 minutes).
        self._tokens[installation_id] = (data["token"], time.time() + 55 * 60)
        return data["token"]

    def post_pr_comment(
        self,
        repo_full_name: str,
        pr_number: int,
        installation_id: str,
        body: str,
    ) -> bool:
        """Post an issue comment on the PR. Returns whether it succeeded."""
        try:
            token = self.installation_token(installation_id)
            response = httpx.post(
                f"{_API}/repos/{repo_full_name}/issues/{pr_number}/comments",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Accept": "application/vnd.github+json",
                    "User-Agent": "shipyard-reviewbot",
                },
                json={"body": body},
                timeout=30.0,
            )
            if response.status_code >= 300:
                logger.warning(
                    "PR comment post failed: HTTP %s for %s#%s",
                    response.status_code,
                    repo_full_name,
                    pr_number,
                )
                return False
            return True
        except httpx.HTTPError as error:
            logger.warning("PR comment post errored for %s#%s: %s", repo_full_name, pr_number, error)
            return False
